lib/_tls_wrap.js

```diff
@@ -37,211 +37,227 @@ const tls_wrap = process.binding('tls_wrap');
-  this._finishInit();
+
+  const owner = this.owner;
+
+  // `newSession` callback wasn't called yet
+  if (owner._newSessionPending) {
+    owner._securePending = true;
+    return;
+  }
+
+  owner._finishInit();
 }
 
 
-function loadSession(self, hello, cb) {
+function loadSession(hello) {
+  const owner = this.owner;
+
   var once = false;
   function onSession(err, session) {
     if (once)
-      return cb(new errors.Error('ERR_MULTIPLE_CALLBACK'));
+      return owner.destroy(new errors.Error('ERR_MULTIPLE_CALLBACK'));
     once = true;
 
     if (err)
-      return cb(err);
+      return owner.destroy(err);
 
-    if (!self._handle)
-      return cb(new errors.Error('ERR_SOCKET_CLOSED'));
+    if (owner._handle === null)
+      return owner.destroy(new errors.Error('ERR_SOCKET_CLOSED'));
 
-    self._handle.loadSession(session);
-    cb(null);
+    owner._handle.loadSession(session);
+    owner._handle.endParser();
   }
 
   if (hello.sessionId.length <= 0 ||
       hello.tlsTicket ||
-      self.server &&
-      !self.server.emit('resumeSession', hello.sessionId, onSession)) {
-    cb(null);
+      owner.server &&
+      !owner.server.emit('resumeSession', hello.sessionId, onSession)) {
+    owner._handle.endParser();
   }
 }
 
 
-function loadSNI(self, servername, cb) {
-  if (!servername || !self._SNICallback)
-    return cb(null);
+function loadSNI(info) {
+  const owner = this.owner;
+  const servername = info.servername;
+  if (!servername || !owner._SNICallback)
+    return requestOCSP(owner, info);
 
-  var once = false;
-  self._SNICallback(servername, function(err, context) {
+  let once = false;
+  owner._SNICallback(servername, (err, context) => {
     if (once)
-      return cb(new errors.Error('ERR_MULTIPLE_CALLBACK'));
+      return owner.destroy(new errors.Error('ERR_MULTIPLE_CALLBACK'));
     once = true;
 
     if (err)
-      return cb(err);
+      return owner.destroy(err);
 
-    if (!self._handle)
-      return cb(new errors.Error('ERR_SOCKET_CLOSED'));
+    if (owner._handle === null)
+      return owner.destroy(new errors.Error('ERR_SOCKET_CLOSED'));
 
     // TODO(indutny): eventually disallow raw `SecureContext`
     if (context)
-      self._handle.sni_context = context.context || context;
+      owner._handle.sni_context = context.context || context;
 
-    cb(null, self._handle.sni_context);
+    requestOCSP(owner, info);
   });
 }
 
 
-function requestOCSP(self, hello, ctx, cb) {
-  if (!hello.OCSPRequest || !self.server)
-    return cb(null);
+function requestOCSP(socket, info) {
+  if (!info.OCSPRequest || !socket.server)
+    return requestOCSPDone(socket);
+
+  let ctx = socket._handle.sni_context;
 
   if (!ctx)
-    ctx = self.server._sharedCreds;
+    ctx = socket.server._sharedCreds;
 
   // TLS socket is using a `net.Server` instead of a tls.TLSServer.
   // Some TLS properties like `server._sharedCreds` will not be present
   if (!ctx)
-    return cb(null);
+    return requestOCSPDone(socket);
```

Comment:
This could still be combined with the upper `if (!ctx)` statement. It might also be interesting to see what `socket._handle.sni_context` and `socket.server._sharedCreds` returns. If we know it is either a object or null / undefined, we could make the check explicit (e.g. `if (ctx === null)`).
```js
if (!ctx) {
  ctx = socket.server._sharedCreds;
  if (!ctx)
    return requestOCSPDone(socket);
}
```
Writing more explicit JS does make a nice difference with Turbofan.

Reply:
Yeah, I don't think we're guaranteed them to be anything in particular because user modules do have access and TLS has previously been broken by making these assumptions elsewhere. (Can't find the PR right now but can link later.)

I will update to make it a nested if though.